src/Workspaces/Core/Portable/Workspace/Solution/CachedSkeletonReferences.cs

```diff
@@ -4,246 +4,407 @@
+            if (referenceSet != null)
+                return referenceSet;
 
-            // if the caller is requiring a particular semantic version, it much match what we have cached.
-            if (version != null && version != _version)
-                return null;
+            // okay, we don't have one. so create one now.
+            ITemporaryStreamStorage? storage;
 
-            return _skeletonReferenceSet;
+            using (await _emitGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            {
+                // after taking the gate, another thread may have succeeded.  See if we can use their version if so:
+                referenceSet = TryGetExistingReferenceSet(version, compilation);
+                if (referenceSet != null)
+                    return referenceSet;
+
+                storage = TryCreateMetadataStorage(workspace, compilation, cancellationToken);
+            }
+
+            if (storage == null)
+            {
+                // unfortunately, we couldn't create one. see if we have one from previous compilation., it might be
+                // out-of-date big time, but better than nothing.
+                referenceSet = TryGetReferenceSet(version: null);
+                if (referenceSet != null)
+                {
+                    workspace.LogTestMessage($"We failed to create metadata so we're using the one we just found from an earlier version.");
+                    return referenceSet;
+                }
+            }
+
+            return new SkeletonReferenceSet(storage, compilation.AssemblyName, new DeferredDocumentationProvider(compilation));
         }
-    }
 
-    /// <summary>
-    /// Return a metadata reference if we already have a reference-set computed for this particular <paramref name="version"/>.
-    /// If a reference already exists for the provided <paramref name="properties"/>, the same instance will be returned.  Otherwise,
-    /// a fresh instance will be returned.
-    /// </summary>
-    public MetadataReference? TryGetReference(VersionStamp version, MetadataReferenceProperties properties)
-        => TryGetReferenceSet(version)?.GetMetadataReference(properties);
+        private SkeletonReferenceSet? TryGetExistingReferenceSet(VersionStamp version, Compilation compilation)
+        {
+            // first, check if we have a direct mapping from this compilation to a reference set. If so, use it.  This
+            // ensures the same compilations will get same metadata reference.
+            if (s_compilationToReferenceMap.TryGetValue(compilation, out var referenceSet))
+                return referenceSet;
 
-    private sealed class SkeletonReferenceSet
-    {
-        private readonly object _gate = new();
+            // Then see if we already have a reference set for this version.  if so, we're done and can return that.
+            return TryGetReferenceSet(version);
```

Comment:
as above the ordering seems weird.  sometimes `TryGetReferenceSet` is called before the `s_compilationToReferenceMap`